src/persist-types/src/schema.rs

```diff
@@ -460,6 +482,51 @@ mod tests {
         // Regression test for migrating a RelationDesc with no columns
         // (which gets encoded as a NullArray) to a RelationDesc with one
         // nullable column.
-        testcase(Null, struct_([("a", Boolean, true)]), Some(false))
+        testcase(Null, struct_([("a", Boolean, true)]), Some(false));
+
+        // Test that we can migrate the old maparray columns to listarrays if the contents match.
+        testcase(
+            Map(
+                Field::new_struct(
+                    "map_entries",
+                    vec![
+                        Field::new("keys", Utf8, false),
+                        Field::new("values", Boolean, true),
+                    ],
+                    false,
+                )
+                .into(),
+                true,
+            ),
+            List(
+                Field::new_struct(
+                    "map_entries",
+                    vec![
+                        Field::new("keys", Utf8, false),
+                        Field::new("values", Boolean, true),
+                    ],
+                    false,
+                )
+                .into(),
+            ),
+            Some(false),
+        );
+
+        // Recursively migrate list contents
```

Comment:
Just so I understand, are we expecting any non-NoOp List -> List migrations in practice? Or is this more of a "we have the code for map->list may as well use it" sort of thing?

Reply:
I was originally planning to skip the recursive migration, but then I remembered that Parker introduced some changes in in the second round of fixes: https://github.com/MaterializeInc/materialize/commit/c4c372ccee805c61c3d324a7b1313911e9dba123#diff-e1f6a1defca880cd4c3d7c585c3556a1a7f8cfe01fd6ebcfe12712e5116cb3e3R717-R721.

I'm still not 100% sure this arises in practice, but it seemed easier to add code to handle it than to pin that down, especially since we expect to want it eventually.